### PR TITLE
Apply `--add-exports` to Java 16+ services based on manifest entries

### PR DESCRIPTION
**Pull request by carterkozak**
This provides the same functionality as
https://github.com/palantir/gradle-baseline/pull/1944/

We're piggy-backing off the `Add-Exports` manifest property to collect exports required for any given application based on the libraries it uses.

==COMMIT_MSG==
Apply `--add-exports` to Java 16+ services based on manifest entries
==COMMIT_MSG==

### CODE REVIEW

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.stream.Stream;
+import org.gradle.api.JavaVersion;
+import org.gradle.api.Project;
+import org.gradle.api.file.FileCollection;
+
+/**
+ * Utility functionality which mirrors {@code Add-Exports} plumbing from
+ * <a href="https://github.com/palantir/gradle-baseline/pull/1944">gradle-baseline#1944</a>.
+ */
+final class ModuleExports {
```
Comment (fawind) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java:37]: I think it would be nice to to also have a small explanation here instead of just linking to the PR. Something similar to what you added in:

https://github.com/palantir/gradle-baseline/blob/1a052650da4af39680a7dd3f790a5b291086158e/gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineModuleJvmArgs.java#L42-L47

**Review by fawind — APPROVED** [review]
Looks good, I think the code duplication with the gradle-baseline feature is actually not too bad!

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        // --add-exports is unnecessary prior to java 16
+        if (javaVersion.compareTo(JavaVersion.toVersion("16")) < 0) {
+            return ImmutableList.of();
+        }
+
+        return Stream.concat(
+                        classpath.getFiles().stream().flatMap(file -> {
+                            try {
+                                if (file.getName().endsWith(".jar") && file.isFile()) {
+                                    JarFile jar = new JarFile(file);
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java]: driveby: shouldn't JarFiles be in a try-with-resources block?


```java
    public void close() throws IOException {
        if (closeRequested) {
            return;
        }
        closeRequested = true;

        synchronized (this) {
            // Close streams, release their inflaters, release cached inflaters
            // and release zip source
            try {
                res.clean();
            } catch (UncheckedIOException ioe) {
                throw ioe.getCause();
            }
        }
    }
```

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java (hunk 1 above)
Comment (carterkozak) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java:37]: Good call, updated!

File: gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java (hunk 2 above)
Comment (carterkozak) [reply gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/service/tasks/ModuleExports.java]: Good shout, we should have an error-prone rule for this!
